### PR TITLE
Add reshape no copy (reshape  in java ported to c++)

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?

Some aspects of reshape assumed many things about how the java reshape worked.
This adds an explicit op for that behavior to allow proper c++ delegation/abstraction
when reshaping ndarrays.
This adds new dedicatred array options flags for detecting whether an array needs a copy or not to compliment the op.

## How was this patch tested?

Manually with existing tests.
## Quick checklist

The following checklist helps ensure your PR is complete:

- [ X] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.konduit.ai/multi-project/how-to-guides/contribute/eclipse-contributors) page for details
- [ X] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [X ] Created tests for any significant new code additions.
- [X ] Relevant tests for your changes are passing.

### CODE REVIEW

File: libnd4j/include/array/ArrayOptions.h [hunk 1]
```diff
@@ -96,6 +96,9 @@
 //flags for when array has a view or not
 #define ARRAY_IS_VIEW 33554432
 
+#define ARRAY_IS_VIEW 33554432
```
Comment (treo) [inline libnd4j/include/array/ArrayOptions.h]: Looks like a PR breakout bug? The constant is defined twice.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/shape/PaddingUtils.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+import org.nd4j.linalg.api.memory.MemoryWorkspace;
+import org.nd4j.linalg.api.shape.options.ArrayOptionsHelper;
+import org.nd4j.linalg.factory.Nd4j;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class PaddingUtils {
+
+    // ... [Previous methods remain unchanged] ...
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/shape/PaddingUtils.java]: Looks like a co-pilot mistake?

**Review by treo — CHANGES_REQUESTED** [review]
This needs a tiny bit of cleanup.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/shape/PaddingUtils.java (hunk 2 above)
Comment (agibsonccc) [reply nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/shape/PaddingUtils.java]: Just left over. Thanks for flagging!

**Review by treo — APPROVED** [review]

**Pull request merged**